src/users.py

```diff
@@ -0,0 +1,440 @@
+
+    for user in users:
+        if nick is not None and user.nick != nick:
+            continue
+        if ident is not None and user.ident != ident:
+            continue
+        if host is not None and user.host != host:
+            continue
+        if realname is not None and user.realname != realname:
+            continue
+        if account is not None and user.account != account:
+            continue
+
+        if not potential or allow_multiple:
+            potential.append(user)
+        else:
+            raise ValueError("More than one user matches: " +
+                             _arg_msg.format(nick, ident, host, realname, account, allow_bot))
+
+    if not potential and not allow_none:
+        raise KeyError(_arg_msg.format(nick, ident, host, realname, account, allow_bot))
+
+    if allow_multiple:
+        return potential
+
+    if not potential: # allow_none
+        return None
+
+    return potential[0]
+
+def add(cli, *, nick, ident=None, host=None, realname=None, account=None, channels=None, raw_nick=False):
+    """Create a new user, add it to the user list and return it.
+
+    This function takes up to 6 keyword-only arguments (and no positional
+    arguments): nick, ident, host, realname, account and channels.
+    With the exception of the first one, any parameter can be omitted.
+    If a matching user already exists, a ValueError will be raised.
+
+    The raw_nick keyword argument may be set if the nick has not yet
+    been parsed. In that case, ident and host must both be None, and
+    nick must be in the form nick!ident@host.
+
+    """
+
+    if raw_nick:
+        if ident is not None or host is not None:
+            raise ValueError("ident and host need to be None if raw_nick is True")
+        nick, ident, host = parse_rawnick(nick)
+
+    if exists(nick, ident, host, realname, account, allow_multiple=True, allow_bot=True):
+        raise ValueError("User already exists: " + _arg_msg.format(nick, ident, host, realname, account, True))
+
+    if channels is None:
+        channels = {}
+    else:
+        channels = dict(channels)
+
+    cls = User
+    if predicate(nick):
+        cls = FakeUser
+
+    new = cls(cli, nick, ident, host, realname, account, channels)
+    all_users.add(new)
+    return new
+
+def exists(*args, allow_none=False, **kwargs):
+    """Return True if a matching user exists.
+
+    Positional and keyword arguments are the same as get(), with the
+    exception that allow_none may not be used (a RuntimeError will be
+    raised in that case).
+
+    """
+
+    if allow_none: # why would you even want to do that?
+        raise RuntimeError("Cannot use allow_none=True with exists()")
+
+    try:
+        get(*args, **kwargs)
+    except (KeyError, ValueError):
+        return False
+
+    return True
+
+_raw_nick_pattern = re.compile(
+
+    r"""
+    \A
+    (?P<nick>  [^!@\s]+ (?=!|$) )? !?
+    (?P<ident> [^!@\s]+         )? @?
+    (?P<host>  \S+ )?
+    \Z
+    """,
```

Comment:
I hope you're not talking about the regex in match_hostmask, because it's impossible to reason about. Having it on multiple lines makes it a lot easier to reason about. I'll do non-greedy though, forgot about that.